data/blog/perfectly-fitting-text-to-container-in-react.md

```diff
@@ -0,0 +1,295 @@
+
+## Performance
+
+Measuring the size of an element via `element.getBoundingClientRect()` causes a [reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) which can be catastrophically bad for performance. Manually blocking the browser paint is _doubly_ worrying. I was pretty suspicious of JavaScript auto-size approach for these reasons, but suspicion alone isn't a good enough reason to not do something.
+
+I wouldn't ship a feature like this without extensive instrumentation. It is _critical_ to know a few things:
+
+1. How fast does the resize iteration run in the wild on real computers?
+2. How often and why does the resize iteration fail to converge?
+
+### Local Profiling
+
+I did some local profiling just as a sanity check, and was pretty comfortable with the results. I did so little of this I won't even bother explaining it. It seemed fine on my machine even with CPU throttling way up. I have very little interest in my MacBook's performance.
+
+### Real User Telemetry
+
+Obviously, I used Sentry for this.
+
+```jsx
+const span = Sentry.startInactiveSpan({
+  op: 'function',
+  name: 'AutoSizedText.iterate',
+  onlyIfParent: true,
+});
+
+... run resize algorithm
+
+span.setAttribute('widthDifference', widthDifference);
+span.setAttribute('heightDifference', heightDifference);
+span.setAttribute('iterationCount', iterationCount);
+span.end();
+```
+
+This creates a transaction for every iteration cycle, which notes the duration, the final difference in the child/parent size, and how many iterations it took to get there. That lets me plot the p50, p95, and p99 of the iteration duration as well as track any runs that exceeded the iteration maximum.
+
+Aside: This is out of scope, but in Sentry a query like `p50(transaction.duration)` filtered to `transaction:AutoSizedText.iterate` will return what I need.
+
+The results were fascinating! The instrumentation revealed some interesting performance numbers, and also a few bugs.
+
+- p50 is 0.90ms
+- p95 is 6.00ms
+- p99 is 17.58ms
+- p100 is 75.90ms
+- the max iteration count is around 10
+
+This performance is acceptable to me. The resize happens fast enough to be imperceptible (faster than 100ms) is all cases. Moreover, since the data for the widgets loads async via HTTP, the UI has to watch for a `fetch()` call before the resize anyway, and compared to the `fetch()` duration, this resize duration is meaningless. Lastly, the `fetch()` calls all finish at different times, which amortizes the cost of the resize since they don't have to happen for all the elements at once.
+
+In the case of a page resize, they _do_ have to run all at once, but in my opinion that experience is still acceptable. Waiting a few hundred milliseconds on page resize is normal, and there are other more expensive operations running during a resize anyway.
+
+Success! I can ship, and reduce the max iteration count to 20, just to be generous.
```

Comment:
If you have a binary search, why do you need a max iteration count? You can break if high - low < error margin as that will indicate that you are looking for a size that is below the margin. Also, a 20 max iteration count for a Ologn algorithm seems excessive, it means you are effectively supporting a search from 5px to ~1m pixels. I'm fairly sure you can remove the iteration count from the API and just rely on the high-low margin trigger

Reply:
Expanded the section below to explain this. The limit is there to prevent runaway measurements, which happen for subtle reasons on client machines. Just a safety measure 🐛

Reply:
I guess my question is, why do you worry about runaway measurements and why do you think they wouldnt be handled by that breakpoint? 

The reason I ask is because we have very similar logic (in fact, almost the same thing) in profiling that runs when we try and figure out if the text fits inside a rectangle on a flamegraph where we use a check like that to ensure we are not trying to measure sub px precision and we've never had any issues :)

Reply:
> why do you worry about runaway measurements and why do you think they wouldnt be handled by that breakpoint?

I'm worried because I caused a bunch of runway measurements in production without realizing it, by messing up the comparison that triggers the breakpoint. I eliminated all the ones that I saw, so I'm pretty confident it's fine, but I also want it to be resistant to future bugs. It's not much harm to add the iteration limit, and I'm risk-averse

Reply:
Understood! I think its fine to keep a check like that if it helps you feel safer about at runtime and ensure it doesnt explode. My preference is to do the opposite, but I dont think that makes any of the approaches better or worse.